core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregatorUtils.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.batch.aggregator;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.aggregation.DeduplicateAggregationDefinition;
+import io.cdap.cdap.etl.api.relational.Expression;
+import io.cdap.cdap.etl.api.relational.ExpressionFactory;
+import io.cdap.cdap.etl.api.relational.InvalidRelation;
+import io.cdap.cdap.etl.api.relational.Relation;
+import io.cdap.cdap.etl.api.relational.RelationalTranformContext;
+import io.cdap.cdap.etl.api.relational.StringExpressionFactoryType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Utility class for DedupAggregator.
+ */
+public class DedupAggregatorUtils {
+
+    public static Relation dedupTransform(RelationalTranformContext relationalTranformContext, Relation relation,
+                                          Schema schema, DedupConfig.DedupFunctionInfo filterFunction,
+                                          List<String> uniqueFields) {
+        Optional<ExpressionFactory<String>> expressionFactory = relationalTranformContext.getEngine().
+                getExpressionFactory(StringExpressionFactoryType.SQL);
+        if (!expressionFactory.isPresent()) {
+            return new InvalidRelation("expression factory is not present");
+        }
+        DeduplicateAggregationDefinition.FilterFunction aggFilterFunction = null;
+        if (filterFunction.getFunction().equals(DedupConfig.Function.MAX)) {
+            aggFilterFunction = DeduplicateAggregationDefinition.FilterFunction.MAX;
+        } else if (filterFunction.getFunction().equals(DedupConfig.Function.MIN)) {
+            aggFilterFunction = DeduplicateAggregationDefinition.FilterFunction.MIN;
+        } else {
+            return new InvalidRelation("filterOperation needs to either min or max");
+        }
+
+        Map<String, Expression> selectExpressions = new HashMap<>();
+        List<Expression> dedupExpressions = new ArrayList<>();
+        ExpressionFactory<String> stringExpressionFactory = expressionFactory.get();
+
+        Expression filterExpression = stringExpressionFactory.compile(filterFunction.getField());
+
+        for (Schema.Field field : schema.getFields()) {
+            selectExpressions.put(field.getName(), stringExpressionFactory.compile(field.getName()));
```

Comment:
Please ensure you get properly escaped name by calling io.cdap.cdap.etl.api.relational.ExpressionFactory#getQualifiedColumnName

Reply:
If we do this, we need to check if the engine supports this capability.

Reply:
so just to confirm, we need this:

```
selectExpressions.put(field.getName(), stringExpressionFactory.compile(stringExpressionFactory.getQualifiedColumnName(relation, field.getName())))
```

Reply:
Correct. With the recent discussion though, no "compile" would be needed as `getQualifiedColumnName` would return an `Expression`

Reply:
yes, will make that change once that PR is merged.

Reply:
I would create a function to handle this by checking capabilities exposed by the engine.

```
private String getColumnName(Relation rel,  String name) {
  if (expressionFactory.getCapabilities().contains(CoreExpressionCapabilities.CAN_GET_QUALIFIED_COLUMN_NAME)) {
    return expressionFactory.getQualifiedColumnName(relation, name)
  } else {
    return name;
}
```